From 4041cb5d1c0630ad4f8f9127e138edf20c66477a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jes=C3=BAs=20Guitarte?= <jguitarte@data-modul.com>
Date: Thu, 4 Apr 2013 11:25:40 +0200
Subject: [PATCH 40/40] add flip (vertical and horizontal) options on gstreamer
 filter

From SFR: I have rework this patch somewhat. The patch has changed the
formating style of some lines in the booth files. This was now reverted
and the patch is now a atomic commit only with changes for the
flip-option.

Rework-by: Silvio F <silvio.fricke@gmail.com>
---
 sys/v4l2/gstv4l2filter.c | 63 ++++++++++++++++++++++++++++++++++++++++++++++++
 sys/v4l2/gstv4l2object.h |  3 +++
 2 files changed, 66 insertions(+)

diff --git a/sys/v4l2/gstv4l2filter.c b/sys/v4l2/gstv4l2filter.c
index e0f585a..5fe2690 100644
--- a/sys/v4l2/gstv4l2filter.c
+++ b/sys/v4l2/gstv4l2filter.c
@@ -38,6 +38,35 @@ GST_DEBUG_CATEGORY (v4l2filter_debug);
 #define PROP_DEF_ALWAYS_COPY        TRUE
 #define DEFAULT_PROP_DEVICE   "/dev/video0"
 
+static void
+gst_set_properties (GstV4l2Filter * v4l2filter)
+{
+  struct v4l2_control vflip, hflip;
+  gint fd = v4l2filter->src_v4l2object->video_fd;
+
+  if ((v4l2filter->sink_v4l2object->vflip.value == 1) ||
+      (v4l2filter->sink_v4l2object->vflip.value == 0)) {
+    vflip.id = V4L2_CID_VFLIP;
+    vflip.value = v4l2filter->sink_v4l2object->vflip.value;
+    if (v4l2_ioctl (fd, VIDIOC_S_CTRL, &vflip) < 0)
+      GST_WARNING_OBJECT (v4l2filter->sink_v4l2object, "VIDIOC_S_CTRL failed");
+
+  }
+  if ((v4l2filter->sink_v4l2object->hflip.value == 1) ||
+      (v4l2filter->sink_v4l2object->hflip.value == 0)) {
+    hflip.id = V4L2_CID_HFLIP;
+    hflip.value = v4l2filter->sink_v4l2object->hflip.value;
+    if (v4l2_ioctl (fd, VIDIOC_S_CTRL, &hflip) < 0)
+      GST_WARNING_OBJECT (v4l2filter->sink_v4l2object, "VIDIOC_S_CTRL failed");
+  }
+}
+
+enum
+{
+  PROP_VFLIP = 0x81,
+  PROP_HFLIP = 0x82
+};
+
 #undef GST_VIDEO_SIZE_RANGE
 #define GST_VIDEO_SIZE_RANGE "(int) [ 1, 32767]"
 
@@ -103,6 +132,17 @@ gst_v4l2filter_class_init (GstV4l2FilterClass * klass)
   gst_v4l2_object_install_properties_helper (gobject_class,
       DEFAULT_PROP_DEVICE);
 
+  g_object_class_install_property (gobject_class, PROP_VFLIP,
+      g_param_spec_int ("vflip", "Vertical Flip",
+          "When this property is 1 a vertical flip is applied",
+          0, 1, 0, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_HFLIP,
+      g_param_spec_int ("hflip",
+          "Horizontal Flip",
+          "When this property is 1 a horizontal flip is applied",
+          0, 1, 0, G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
+
   gst_element_class_set_details_simple (element_class,
       "V4l2 Video Filter", "Filter/Converter/Video/Scaler",
       "filters Videos", "Michael Olbrich <m.olbrich@pengutronix.de>");
@@ -240,13 +280,22 @@ gst_v4l2filter_set_property (GObject * object, guint prop_id,
 {
   GstV4l2Filter *v4l2filter = GST_V4L2FILTER (object);
 
+  struct v4l2_control vflip, hflip;
+
   if (!gst_v4l2_object_set_property_helper (v4l2filter->sink_v4l2object,
           prop_id, value, pspec)) {
     switch (prop_id) {
+      case PROP_HFLIP:
+        v4l2filter->sink_v4l2object->hflip.value = g_value_get_int (value);
+        break;
+      case PROP_VFLIP:
+        v4l2filter->sink_v4l2object->vflip.value = g_value_get_int (value);
+        break;
       default:
         G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
         break;
     }
+    gst_set_properties (v4l2filter);
   }
 }
 
@@ -259,6 +308,14 @@ gst_v4l2filter_get_property (GObject * object, guint prop_id, GValue * value,
   if (!gst_v4l2_object_get_property_helper (v4l2filter->sink_v4l2object,
           prop_id, value, pspec)) {
     switch (prop_id) {
+      case PROP_HFLIP:
+        value = v4l2filter->sink_v4l2object->hflip.value;
+        break;
+
+      case PROP_VFLIP:
+        value = v4l2filter->sink_v4l2object->vflip.value;
+        break;
+
       default:
         G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
         break;
@@ -458,6 +515,10 @@ gst_v4l2filter_sink_setcaps (GstPad * pad, GstCaps * caps)
   if (!gst_v4l2filter_src_update_caps (v4l2filter, caps))
     return FALSE;
 
+
+  printf ("set properties in: gst_v4l2filter_sink_setcaps");
+  gst_set_properties (v4l2filter);
+
   return TRUE;
 }
 
@@ -520,6 +581,8 @@ gst_v4l2filter_src_setcaps (GstPad * pad, GstCaps * caps)
 
   gst_pad_push_event (pad, gst_event_new_caps (caps));
 
+  printf ("set properties in: gst_v4l2filter_src_setcaps");
+  gst_set_properties (v4l2filter);
   return TRUE;
 }
 
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index 3fa6d85..a7f0793 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -158,6 +158,9 @@ struct _GstV4l2Object {
   GstV4l2GetInOutFunction  get_in_out_func;
   GstV4l2SetInOutFunction  set_in_out_func;
   GstV4l2UpdateFpsFunction update_fps_func;
+
+  struct v4l2_control vflip;
+  struct v4l2_control hflip;
 };
 
 struct _GstV4l2ObjectClassHelper {
-- 
1.8.4.2

